### ISSUE TITLE
Thoughts on `TeleopDriveCommand` in `arcade-drive` branch

### ISSUE THREAD

**Issue by wxb1ank** [BadRobots1014/BadRobot2022#1]
I really like the changes to the `arcade-drive` branch so far. Here are my thoughts/suggestions:

https://github.com/BadRobots1014/BadRobot2022/blob/e17d6c63a708c50ddbf81401cdba2afac84bd2ae/src/main/java/frc/robot/commands/TeleopDriveCommand.java#L54-L61
I don't think this method should be public if the purpose of `TeleopDriveCommand` is to select a strategy based on the current joystick position—in that case, I think `TeleopDriveCommand` should be responsible for calling `getNextDriveStrategy` and `setStrategy`, and the only public interface would be `TeleopDriveCommand::execute`.

However, if we decide that there should only be one drivetrain command, and anchoring, driving straight, arcade-driving, etc. are only enabled through `TeleopDriveCommand`, then this makes sense because it may be desirable to manually enable a specific strategy if the driver is holding a special button.

https://github.com/BadRobots1014/BadRobot2022/blob/e17d6c63a708c50ddbf81401cdba2afac84bd2ae/src/main/java/frc/robot/commands/TeleopDriveCommand.java#L70-L89
In its current form, this function always evaluates all of the conditions. Also, `m_xSource::get` and `m_ySource::get` are called multiple times.

This could be rewritten as:
```java
final double x = m_xSource.get();
final double y = m_ySource.get();

if (DriveStrategyResolutionUtil.isInputForDriveStraight(x, y)) {
    return this.m_driveStraightStrategy;
} else if (DriveStrategyResolutionUtil.isInputForPivotTurn(x, y)) {
    return this.m_pivotTurnStrategy;
} else if (DriveStrategyResolutionUtil.isInputForHoldPlace(x, y)) {
    return this.m_holdPlaceStrategy;
} else {
    /*
    * For now, let's just default to ArcadeDriveStrategy.
    */
    return this.m_arcadeStrategy;
}
```
Another issue is a (mostly insignificant) race condition that, AFAICT, exists because, for every execution of a drive strategy, the joystick's position is polled twice: once in `TeleopDriveCommand::getNextDriveStrategy`, and once in `TeleopDriveCommand::execute`. For example:
1. Client code calls `getNextDriveStrategy`, which says that the next strategy should be to pivot turn. Client code sets this strategy with `setStrategy`.
2. The driver releases the joystick back to the origin.
3. The command scheduler calls `TeleopDriveCommand::execute`, which polls the joystick's position—which is now (0, 0)—and passes it to `PivotTurnStrategy::execute`.
4. `PivotTurnStrategy` does not rotate the robot at all because the X-coordinate is 0.

https://github.com/BadRobots1014/BadRobot2022/blob/e17d6c63a708c50ddbf81401cdba2afac84bd2ae/src/main/java/frc/robot/commands/drive/ArcadeDriveStrategy.java#L33-L44
I don't think that strategies should be able to modify the `TeleopDriveCommand` directly; in my opinion, this creates an unobvious bidirectional relationship. Rather, I believe `TeleopDriveCommand` should be solely responsible for queuing strategies. Also, this snippet must be duplicated in every strategy; copy-and-pasting could lead to subtle errors.

**Comment by vchen04** [comment]
We could make it so that a `DriveStrategyContext` such as `TeleopDriveCommand` is responsible for choosing its own strategy, and I believe that it may be better to structure it this way. This way, the `DriveStrategy` classes are only responsible for implementing the algorithm to handle joystick inputs in a specific way, and are not also responsible for managing the state of the `DriveStrategyContext`. An advantage to this approach is that `DriveStrategy` code can be reused in any `Command` that uses a `DriveStrategy` without regard to if and how said `Command` implements state.

This would make it so we do not need to use include the if statement in the following snippet:

https://github.com/BadRobots1014/BadRobot2022/blob/e17d6c63a708c50ddbf81401cdba2afac84bd2ae/src/main/java/frc/robot/commands/drive/ArcadeDriveStrategy.java#L33-L44

One way that a `TeleopDriveCommand` could be rewritten to do this is by checking if `this.m_strategy` and `this.getNextDriveStrategy()` are different objects, and then executing a state change through `this::setStrategy`.

**Comment by wxb1ank** [comment]
> We could make it so that a `DriveStrategyContext` such as `TeleopDriveCommand` is responsible for choosing its own strategy

I like this approach. I think it provides clear logical separation between strategy selection and algorithm.

> One way that a `TeleopDriveCommand` could be rewritten to do this is by checking if `this.m_strategy` and `this.getNextDriveStrategy()` are different objects, and then executing a state change through `this::setStrategy`.

If we do this (and I'm not sure what alternatives there are), then we could do `newStrategy.getClass() == oldStrategy.getClass()` to check for class equivalence. Otherwise, `TeleopDriveCommand::getNextDriveStrategy` will probably either need to return some kind of unique strategy identifier (e.g., enum) or a new strategy object with a static unique identifier field; I dislike the latter approach as it injects implementation details of `TeleopDriveCommand` into `DriveStrategy`.

**Comment by vchen04** [comment]
> 

We could also be lazy. If we initialize the `DriveStrategy` objects in `TeleopDriveCommand` once and only once, then we could just check for reference equality between `this.getNextDriveStrategy()` and `this.m_strategy`.

**Comment by wxb1ank** [comment]
> We could also be lazy. If we initialize the `DriveStrategy` objects in `TeleopDriveCommand` once and only once, then we could just check for reference equality between `this.getNextDriveStrategy()` and `this.m_strategy`.

I've implemented this in the `arcade-drive-testing` branch (which has also been merged into `master`).

Also, it is unnecessary for drive strategies to have a reference back to `DriveTrainSubsystem`; the `DriveStrategy::execute` method now returns a `WheelSpeeds` class (containing the left and right motor power values) which `TeleopDriveCommand` feeds into `DriveTrainSubsystem::tankDrive`.

At this point, there are two more improvements I'd like to make:

1. I would ideally like to eliminate `DriveStrategy::reset` and move the resetting code into the constructors of drive strategies. This is feasible because the strategies no longer contain the PID controllers—I moved them into `GyroSubsystem`—so it should be fairly cheap to construct a new strategy object every time one is selected. My main motivation for doing this is, we would no longer need to store all four drive strategies as instance variables in `TeleopDriveCommand`. However, that also means we can no longer use reference equality for determining if the next strategy is the same as the last one.
2. The mentors have expressed that `DriveStrategy::execute` should be refactored somehow such that drive strategies no longer have to accept both `x` and `y` arguments from `TeleopDriveCommand`. I think the main benefit is to clear confusion around why e.g. `AnchorStrategy::execute` requires `x` and `y` arguments and yet does nothing with them. Personally, I do not think this is a major problem, and there isn't really any performance loss with the current implementation as the joystick position isn't re-polled when executing a drive strategy; `TeleopDriveCommand::execute` needs to poll the position to select the next strategy, anyway. And, I think devising an alternative to the current implementation will be tricky; if each strategy has a different `execute()` interface, then `TeleopDriveCommand` can no longer abstract over the different strategies so easily, and I think we will either need to downcast or use quite a few `switch` statements. (Alternatively, we can store the joystick coordinates post-scaling in instance variables and pass providers of them to the constructors of drive strategies that require `x`/`y`, but I'm not sure such a solution is any less worthy of refactoring.) Please let me know if you have any ideas, or if I'm missing something.

**Comment by vchen04** [comment]
> it should be fairly cheap to construct a new strategy object every time one is selected. My main motivation for doing this is, we would no longer need to store all four drive strategies as instance variables in `TeleopDriveCommand`

You're right that it would be fairly cheap to construct a new strategy object every time one is selected. However, I don't see why it's necessarily bad to store references to all four drive strategies as instance variables. Additionally, it might be easier to understand for someone new looking at the code if we do use reference equality.

> The mentors have expressed that `DriveStrategy::execute` should be refactored somehow such that drive strategies no longer have to accept both `x` and `y` arguments from `TeleopDriveCommand`.

One thing you could do is have the concrete `DriveStrategy` classes that need joystick input store references to `Supplier<Double>`s, and have these references initialized in the constructors. How these objects are constructed is not defined in the `DriveStrategy` interface, so it wouldn't necessarily be bad form (at least as far as I can tell... @kanno41 ?) to do whatever we need to in order to set up these objects.

**Comment by vchen04** [comment]
> Additionally, it might be easier to understand for someone new looking at the code if we do use reference equality.

Another approach that might work is to override the `equals` method to conform to our definition of equality, though we would not be able to implement this in the `DriveStrategy` interface. If you really wanted to go down this route, we could create an abstract class that overrides the `equals` method using a layered implementation (it uses methods we defined in our interface to do work to check for equality, so the implementation of this `equals` method is independent of the implementation of each concrete `DriveStrategy`, http://web.cse.ohio-state.edu/software/2231/web-sw2/extras/slides/03.Abstract-Classes.pdf), or write a static method in the interface (we can do that in Java), also with a layered implementation, to check for equality.

**Comment by wxb1ank** [comment]
### "Improvement" 1: eliminating `DriveStrategy::reset` and finding an alternative to testing reference equality

> However, I don't see why it's necessarily bad to store references to all four drive strategies as instance variables. Additionally, it might be easier to understand for someone new looking at the code if we do use reference equality.

It's not necessarily bad, and you could be right that it would be easier to understand. I just prefer less state in general, and especially when that state is actually acting as a disjoint union, which opens the door for accidental mutation of strategies not currently in-use.

> Another approach that might work is to override the equals method to conform to our definition of equality

I suppose I would have to see an example to know what this looks like (the OSU link is 404'ing for me), but if you want to do this, I personally would create a new method rather than override `Object::equals`, which AFAIK is meant for object equality rather than class equality (i.e., comparing state vs. comparing type). An alternative I just thought of, but would not recommend, is testing for `DriveStrategy::getName` equality.

My original idea was to store a strategy "identifier" (e.g., enum) and a type-erased `DriveStrategy` object in `m_strategy` using some container type, and to test for equality using the identifier and call the drive strategy methods through the type-erased object (I'm trying to emulate a tagged/disjoint union, but I don't come from a Java background so this is probably unidiomatic). Of course, that is pointless if we decide the `DriveStrategy::execute` interface should be different for each strategy (although we might pass `Supplier<Double>`s as you mentioned), and this solution is also ugly in general:

1. You now must uphold the invariant that the identifier and associated `DriveStrategy` object correspond to each other; and
2. It attempts to reconcile two conflicting goals.

I want to explain that second point. In its current form, the presented problem is impossible because we require drive strategies to have the conflicting traits of being uniquely identifiable while sharing a common interface. Using reference equality and `DriveStrategy::getName` sidestep this problem as the unique identifier is part of the interface (e.g., `Object::equals`, `DriveStrategy::getName`). Therefore, there are really only two solutions:

1. Implement the unique identifer in the interface (e.g., with a method); or
2. Create the unique identifier outside the interface (e.g., with an enum) and assign it to each object.

The first solution is what we've currently implemented, and it is consistent with our decision to unite the drive strategies under the common `DriveStrategy` interface.

To be similarly consistent, the second solution would **not** use a common interface (i.e., no `DriveStrategy`), and might look something like the following:

*ArcadeDriveStrategy.java*
```java
public class ArcadeDriveStrategy {
    // ...

    // No 'prepare' method needed.

    public static WheelSpeeds getWheelSpeeds(
        final double linearPower,
        final double rotationalPower
    ) {
        // ...
    }
}
```

*AnchorStrategy.java*
```java
public class AnchorStrategy {
    // ...

    public static void prepare(final GyroSubsystem gyro) {
        // This also zeroes the rotational PID setpoint.
        gyro.zeroYaw();
    }

    public static WheelSpeeds getWheelSpeeds() {
        // ...
    }
}
```

*TeleopDriveCommand.java*
```java
public class TeleopDriveCommand {
    private final DriveTrainSubsystem drive;
    private final GyroSubsystem gyro;
    private Optional<Strategy> currentStrategy;

    // ...

    @Override
    public void execute() {
        final double x, y;
        // ...
        final Strategy nextStrategy = this.selectNextStrategy(x, y);
        if (nextStrategy != this.currentStrategy) {
            // The next strategy is different from the last one, so we must prepare it first.
            switch (nextStrategy) {
                case ANCHOR:
                    AnchorStrategy.prepare(this.gyro);
                    break;
                default:
                    // No preparation required.
            }
        }

        final WheelSpeeds wheelSpeeds = this.getWheelSpeeds(nextStrategy, x, y);
        this.drive.tankDrive(wheelSpeeds.left, wheelSpeeds.right);
    }

    private static enum Strategy {
        ARCADE_DRIVE,
        ANCHOR
    }

    private static Strategy selectNextStrategy(final double x, final double y) {
        if (/* ... */) {
            return Strategy.ANCHOR;
        }

        return Strategy.ARCADE_DRIVE;
    }

    private static WheelSpeeds getWheelSpeeds(Strategy strategy, final double x, final double y) {
        switch (nextStrategy) {
            case ARCADE_DRIVE:
                // Y is the linear power, and X is the rotational power.
                return ArcadeDriveStrategy.getWheelSpeeds(y, x);
            case ANCHOR:
                return AnchorStrategy.getWheelSpeeds();
        }
    }
}
```

As you can see, once we remove the requirement that every drive strategy implements a common interface, each drive strategy may become stateless; each drive strategy class is merely a collection of functions now.

What do you think of this approach? There's no object allocation, and I think the code is comprehensible. As a bonus, it addresses Improvement 2.

### Improvement 2: Refactoring `DriveStrategy::execute`

> One thing you could do is have the concrete `DriveStrategy` classes that need joystick input store references to `Supplier<Double>`s, and have these references initialized in the constructors. How these objects are constructed is not defined in the `DriveStrategy` interface, so it wouldn't necessarily be bad form (at least as far as I can tell... @kanno41 ?) to do whatever we need to in order to set up these objects.

I thought of this, and I did something similar in my evil arcade-drive implementation with `m_cachedX` and `m_cachedY` instance variables. Again, that's a fine solution in this case, but I just dislike it in principle as there is now an indirect relationship between the joystick position polled in `TeleopDriveCommand::execute` and that received in the drive strategy methods; you must now uphold the invariant that the cached position is always equal to the original polled position (e.g., it's now possible to forget to update the cached position, and that wasn't possible before as those arguments were necessary to call `DriveStrategy::execute`).

**Comment by BowlesCR** [comment]
Why not compare based on `.getClass()` ?
Overriding the `equals()` method is the accepted way to do such a thing.

**Comment by vchen04** [comment]
> when that state is actually acting as a disjoint union
I have no idea what this means...

Additionally, the entire point of having a common interface is to improve modularity, so we can drop these strategies in wherever we need to (autonomous 👀), so I'd prefer not to throw that out.

I don't know if there's a good solution to the `DriveStrategy::execute` problem now. Caching `x` and `y` inputs from the joystick seems okay as long as you're careful about it. With caching, these values can be controlled by `TeleopDriveCommand` and fed into a `DriveStrategy`, so there is a more linear flow of information. But, I'm going to step back from this because I have a headache and a mountain of homework. If anyone else could chime in, that would be great.

**Comment by wxb1ank** [comment]
> Why not compare based on `.getClass()` ?

That works.

>> when that state is actually acting as a disjoint union
> I have no idea what this means...

Sorry, I could've explained that better. I mean that, although `TeleopDriveCommand` has four drive strategy instance variables, it is only ever supposed to access one at a time. Thus, what we really want is a disjoint/tagged union (or variant or coproduct), which would enforce that at the type-/language-level. But because Java AFAIK doesn't have those, we must uphold this invariant ourselves. (Now, I've since realized that, in practice, this shouldn't really matter as we currently call `DriveStrategy::reset` when starting a new strategy anyway.)

> Additionally, the entire point of having a common interface is to improve modularity, so we can drop these strategies in wherever we need to (autonomous 👀), so I'd prefer not to throw that out.

I would agree, and if that's a problem, then we should keep the current approach. I will say, though, that I think the impact would be minimal. That is to say, in `DriveStrategyCommand`, a factory pattern is used to create each kind of drive strategy. If the drive strategies were to no longer share a common interface, then the only real change would be moving these factory methods into separate `ArcadeDriveCommand`, `AnchorCommand`, etc. classes that are simply wrappers for the underlying strategies. The asynchronous code will probably be composing commands together rather than strategies (although we could implement an interface to do the latter), so I think the impact of this would basically be limited to `DriveStrategyCommand`.

I don't mean to block progress on or hold up more important things, so we can move on and keep the current implementation as-is if you'd like, and maybe return to this later. Also please let me know if I am being too abrasive or anything.

**Comment by vchen04** [comment]
Don't worry too much about being too abrasive—it stimulates good conversation, though I am feeling muddled from thinking about this topic too much. From what information I've been able to gather, the current implementation works, so it's fine for now, and refactoring is a later issue (if we decide to deal with it, depending on how much gets put on our plate as the season progresses). There is one thing that I would prefer to have changed though, and I've brought it up in the other issue/discussion (#3). The main reason for this is so that we can use kinematics and odometry in the future.
